### PR TITLE
Triage some SpotBugs violations

### PR DESCRIPTION
**Pull request by basil**
This PR triages the following SpotBugs violations …

- [`DM_NUMBER_CTOR`](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bx-method-invokes-inefficient-number-constructor-use-static-valueof-instead-dm-number-ctor)
- [`DP_CREATE_CLASSLOADER_INSIDE_DO_PRIVILEGED`](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#dp-classloaders-should-only-be-created-inside-doprivileged-block-dp-create-classloader-inside-do-privileged)
- [`MS_MUTABLE_COLLECTION_PKGPROTECT`](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#ms-field-is-a-mutable-collection-which-should-be-package-protected-ms-mutable-collection-pkgprotect)
- [`MS_OOI_PKGPROTECT`](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#ms-field-should-be-moved-out-of-an-interface-and-made-package-protected-ms-ooi-pkgprotect)
- [`MS_PKGPROTECT`](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#ms-field-should-be-package-protected-ms-pkgprotect)
- [`NP_NULL_ON_SOME_PATH`](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#np-possible-null-pointer-dereference-np-null-on-some-path)
- [`RCN_REDUNDANT_NULLCHECK_WOULD_HAVE_BEEN_A_NPE`](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#rcn-nullcheck-of-value-previously-dereferenced-rcn-redundant-nullcheck-would-have-been-a-npe)
- [`SBSC_USE_STRINGBUFFER_CONCATENATION`](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#sbsc-method-concatenates-strings-using-in-a-loop-sbsc-use-stringbuffer-concatenation)
- [`SWL_SLEEP_WITH_LOCK_HELD`](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#swl-method-calls-thread-sleep-with-a-lock-held-swl-sleep-with-lock-held)
- [`URF_UNREAD_FIELD`](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#urf-unread-field-urf-unread-field)
- [`WA_AWAIT_NOT_IN_LOOP`](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#wa-condition-await-not-in-loop-wa-await-not-in-loop)
- [`WA_NOT_IN_LOOP`](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#wa-wait-not-in-loop-wa-not-in-loop)

… and the following FindSecBugs violations …

- [`GROOVY_SHELL`](https://find-sec-bugs.github.io/bugs.htm#GROOVY_SHELL)
- [`WEAK_MESSAGE_DIGEST_SHA1`](https://find-sec-bugs.github.io/bugs.htm#WEAK_MESSAGE_DIGEST_SHA1)
- [`XXE_DOCUMENT`](https://find-sec-bugs.github.io/bugs.htm#XXE_DOCUMENT)
- [`XXE_DTD_TRANSFORM_FACTORY`](https://find-sec-bugs.github.io/bugs.htm#XXE_DTD_TRANSFORM_FACTORY)
- [`XXE_XSLT_TRANSFORM_FACTORY`](https://find-sec-bugs.github.io/bugs.htm#XXE_XSLT_TRANSFORM_FACTORY)

### Proposed changelog entries

N/A

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

File: core/src/main/java/hudson/ExtensionList.java
Comment (basil) [inline core/src/main/java/hudson/ExtensionList.java:331]: Chases away `NP_NULL_ON_SOME_PATH`.

File: core/src/main/java/hudson/ExtensionList.java
Comment (basil) [inline core/src/main/java/hudson/ExtensionList.java:383]: Chases away `NP_NULL_ON_SOME_PATH`.

File: core/src/main/java/hudson/WebAppMain.java
Comment (basil) [inline core/src/main/java/hudson/WebAppMain.java:254]: Was triggering `XXE_DTD_TRANSFORM_FACTORY` and `XXE_XSLT_TRANSFORM_FACTORY`, but the `TransformerFactory` is immediately thrown away rather than used for anything, so this is not actually a problem. Furthermore, this entire block of code is a workaround for [a bug that was fixed in Tomcat 5.5](https://bz.apache.org/bugzilla/show_bug.cgi?id=40895), which was [EOLed in 2012](https://tomcat.apache.org/tomcat-55-eol.html), so it's safe to say we don't need it anymore.

File: core/src/main/java/hudson/model/AbstractItem.java
Comment (basil) [inline core/src/main/java/hudson/model/AbstractItem.java:356]: The sleep is only done in an edge case (retrying when a file rename operation fails on Windows), so I think this is fine overall.

File: core/src/main/java/hudson/model/AutoCompletionCandidates.java
Comment (basil) [inline core/src/main/java/hudson/model/AutoCompletionCandidates.java:112]: Premature optimization if I ever saw it.

File: core/src/main/java/hudson/model/DirectoryBrowserSupport.java
Comment (basil) [inline core/src/main/java/hudson/model/DirectoryBrowserSupport.java:753]: Premature optimization if I ever saw it.

File: core/src/main/java/hudson/model/StringParameterDefinition.java
Comment (basil) [inline core/src/main/java/hudson/model/StringParameterDefinition.java:148]: As of #5275

https://github.com/jenkinsci/jenkins/blob/165d559469c7a58af581931bd1b89e5b9ed4a9af/core/src/main/java/hudson/model/StringParameterValue.java#L51

means that this can no longer be null.

File: core/src/main/java/hudson/security/ACL.java
Comment (basil) [inline core/src/main/java/hudson/security/ACL.java:358]: Triggers `MS_PKGPROTECT` because this is protected (as opposed to package-private) and mutable. SpotBugs has a point. [Nothing else in the ecosystem is using it](https://github.com/search?type=Code&q=user%3Ajenkinsci+AUTOMATIC_SIDS), so might as well restrict this for better hygiene and to keep SpotBugs happy.

File: core/src/main/java/hudson/triggers/SCMTrigger.java
Comment (basil) [inline core/src/main/java/hudson/triggers/SCMTrigger.java:213]: Chases away `NP_NULL_ON_SOME_PATH`.

File: core/src/main/java/hudson/triggers/SCMTrigger.java
Comment (basil) [inline core/src/main/java/hudson/triggers/SCMTrigger.java:695]: Chases away `NP_NULL_ON_SOME_PATH`.

File: core/src/main/java/jenkins/util/DirectedGraph.java
Comment (basil) [inline core/src/main/java/jenkins/util/DirectedGraph.java:80]: Technically unused, but useful when reading the code and following along in Tarjan's algorithm.

File: core/src/main/java/jenkins/util/JSONSignatureValidator.java
Comment (StefanSpieker) [inline core/src/main/java/jenkins/util/JSONSignatureValidator.java:56]: I think we might better drop SHA-1 totally.

File: core/src/main/java/hudson/model/UpdateCenter.java
Comment (basil) [reply core/src/main/java/hudson/model/UpdateCenter.java:1243]: Possibly, though I have no idea what constraints there are regarding compatibility and testing. Better left for a future PR I think.

**Review by StefanSpieker — APPROVED** [review]

File: core/src/main/java/hudson/ClassicPluginStrategy.java
Comment (jglick) [inline core/src/main/java/hudson/ClassicPluginStrategy.java:78]: Can we just add this to the exclude list pending https://github.com/spotbugs/spotbugs/issues/1515?

File: core/src/main/java/hudson/node_monitors/ClockMonitor.java
Comment (jglick) [inline core/src/main/java/hudson/node_monitors/ClockMonitor.java:57]: At some point we should do a pass through things like this and delete any not actually being used, but certainly out of scope here.

File: core/src/main/java/hudson/util/ProcessTree.java
Comment (jglick) [inline core/src/main/java/hudson/util/ProcessTree.java:2245]: Actually Groovy will just ignore access modifiers and let you mutate a `private` field too.

File: core/src/main/java/jenkins/fingerprints/FileFingerprintStorage.java
Comment (jglick) [inline core/src/main/java/jenkins/fingerprints/FileFingerprintStorage.java:329]: https://github.com/spotbugs/spotbugs/issues/756

File: src/spotbugs/spotbugs-excludes.xml
Comment (jglick) [inline src/spotbugs/spotbugs-excludes.xml:223]: BTW I do not follow why entries like this were ever added to begin with. Simpler to just add `@SuppressFBWarnings` to the class in question with a `justification`?

**Review by jglick — APPROVED** [review]

File: src/spotbugs/spotbugs-excludes.xml
Comment (basil) [reply src/spotbugs/spotbugs-excludes.xml:223]: Try removing these entries and adding `@SuppressFBWarnings` annotations for untriaged violations and see how long it takes you.

Comment (jglick) [reply src/spotbugs/spotbugs-excludes.xml:223]: Well, I meant for the entries that have just one or two `<Class>`es.

File: core/src/main/java/hudson/util/ProcessTree.java
Comment (basil) [reply core/src/main/java/hudson/util/ProcessTree.java:2245]: Good point. I checked `usage-in-plugins` and these are unused, so I marked them as package-private and removed the suppression. I also verified that they're still settable from the script console.

File: core/src/main/java/hudson/ClassicPluginStrategy.java
Comment (basil) [reply core/src/main/java/hudson/ClassicPluginStrategy.java:78]: Done

File: src/spotbugs/spotbugs-excludes.xml
Comment (basil) [reply src/spotbugs/spotbugs-excludes.xml:225]: I don't suppose it matters much either way, but I find it easier to triage stuff if the list of stuff to triage is in one place. I did add ~25 `@SuppressFBWarnings(value = "[…]", justification = "TODO needs triage")` annotations to various files where I couldn't easily suppress the warnings from `spotbugs-excludes.xml` (for example, because it was for an anonymous inner class that I couldn't easily refer to), but for the most part I added untriaged stuff to `spotbugs-excludes.xml` to keep the list in one place. I find it convenient to have it organized by violation type and class name in one file. Makes it easy to pick off a section and tackle it one area at a time.

**Review by res0nance — APPROVED** [review]

**Comment by basil** [comment]
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

**Pull request merged**
